sklearn/tree/iforest.py

```diff
@@ -0,0 +1,240 @@
+from sklearn.base import BaseEstimator
+
+
+class IForest(BaseEstimator):
+    """Isolation Forest Algorithm
+
+    Return the anomaly score of each sample in X with the iForest algorithm
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The training data
+
+    n_trees : int
+        The number of trees to build
+
+    psi : int
+        the sub-sampling size used to build a tree
+
+    random_state : integer or numpy.RandomState, optional
+        The generator used to initialize the centers. If an integer is
+        given, it fixes the seed. Defaults to the global numpy random
+        number generator.
+
+
+    Attributes
+    -------
+
+    """
+    def __init__(self, n_trees=100, psi=256, random_state=None):
+        self.n_trees = n_trees
+        self.psi = psi
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
+        self.forest = build_iforest(
+            X, t=self.n_trees, psi=self.psi, random_state=self.random_state)
+        return self
+
+    def predict(self, X):
+        return iforest_score(X, self.forest)
+
+    def decision_function(self, X):
+        return iforest_score(X, self.forest)
+
+    def score(self, X, y):  # return the auROC score
+        scoring = self.predict(X)
+        return roc_auc_score(y, scoring)
+
+    def score2(self, X, y):  # return the both AUC and auPR scores
+        scoring = self.predict(X)
+        precision, recall, thresholds = precision_recall_curve(y, scoring)
+        return auc(recall, precision), roc_auc_score(y, scoring)
+
+
+def iforest_score(X, forest):
+    """Return the anomaly score of each sample in X with the iForest algorithm
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The data to evaluate anomaly score.
+
+    forest : list
+        The forest.
+
+    Returns
+    -------
+    scores : ndarray
+        scores[i] is the anomaly score of the sample X[i, :]
+    """
+    n_samples = X.shape[0]
+    scores = np.zeros(n_samples)
+    for i in range(n_samples):
+        mean_path_length = 0
+        x = X[i, :]
+        for tree in forest:
+            path_length = 0
+            while isinstance(tree, InNode):
+                if tree.compare(x):
+                    tree = tree.right
+                else:
+                    tree = tree.left
+                path_length += 1
+            path_length += c(tree.size)
+            mean_path_length += path_length
+        mean_path_length /= len(forest)
+        scores[i] = np.power(2, - mean_path_length / c(n_samples))
+    return scores
+
+
+def c(n):
+    """ The average path length in a n samples iTree, which is equal to
+        the average path length of an unsuccessful BST search since the
+        latter has the same structure as an isolation tree.
+    """
+    if n <= 1:
+        return 0
+    else:
+        return 2 * H(n) - 2 * (n - 1) / n
+
+
+def H(n):
```

Comment:
Kindly avoid single character variable / method / function names... 

People have a tendency to do `from iforest import *` and could later unknowingly change / assign to `c` / `H` which could cause trouble...